String hierarchy spec testing

johnshi314 opened the pull request.

johnshi314 committed:
Added a test class for StringHierarchySpecTests. It can be found in tests\EStimLibrary.UnitTests\Extensions\SpatialModel\StringHierarchy\StringHierarchySpecTests.cs

johnshi314 committed:
Completed tests for StringHierarchySpec.

lhmcgann commented:
Make sure to edit the PR to merge onto release not main

cld99 commented on tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchySpecTests.cs:
Make sure to update your test progress tracker at the top to reflect that GetHashCode and ToString have been tested. You may also want to test more types of malformed inputs for ParseFullSpec, ParseRegionSpec, and ParseModifierSpec (i.e., extra delimiters and other inputs that do not follow the required convention). Should RegionSetOverlaps and ModifiersAllowOverlap retrun true for partial overlaps (as RegionSetOverlaps seems to operate), or just full overlap of the shorter specification (as ModifiersAllowOverlap seems to operate and as the comments within the methods indicate)? Depending on which is correct, method comments/behavior should probably be adjusted to be consistent. Testing for TryGetOverlap should probably include tests for overlapping modifier sets without overlapping regions just to make sure that it works as expected. Also, shouldn't the expected overlapping region for the second test of TryGetOverlap_ShouldIdentifyOverlapIsPossible just be "region1 | modifier1"? Testing for Equals should probably include tests for different orderings of the regions, since the ordering does matter for those. Likewise for GetHashCode, which you based the testing on for consistency with the Equals method. Other than these small nitpicks, everything looks good!

cld99 reviewed (dismissed):
I left comments on the StringHierarchySpecTest file detailing some small nitpicks, but everything looks to be functioning properly and ready to merge.

lhmcgann committed:
StringHierarchySpec RegionSetOverlap implementation fixed

lhmcgann committed:
IBodyModel validation and StringHierarchySpec overlap reimplementation

lhmcgann commented:
Ok @drmannequin I refactored/reimplemented the TryGetOverlap and associated functions in StringHierarchySpec. It resulted in me moving those methods and that overall functionality to StringHierarchyBodyModel, so the test cases are commented out. You can leave them like that in your test class until you get to testing the body model, but just test all the other StringHierarchySpec methods for now. Lmk if you want to chat, get any clarification, etc.

johnshi314 committed:
IBodyModel validation and StringHierarchySpec overlap reimplementation

lhmcgann committed:
Formatting, comments, simple testing additions

lhmcgann commented on src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchySpec.cs:
```diff
@@ -81,118 +117,70 @@ public static string JoinFullSpec(string[] regionSet, string[] modifierSet)
     /// <returns>True if valid parse, False if not.</returns>
     public static bool TryParseOptionedRegionName(string optionedRegionName,
-        out string baseName, out string options)
+        out string baseName, out string option)
```
This should be "options", e.g., "left index finger" should be valid where "finger" is the region basename and "left index" is the option string for that region. Forgive me if I made these changes making it now incorrect, but (1) it should be reverted back to what is in red. (2) Add a test case or so for this case. (3) Write a method in StringHierarchySpec called TryParseOptions() that uses the OPTIONS_REGION_DELIMITER to split an "options" string as output by TryParseOptionedRegionName, returning a bool (true if any options found, else false) and outputting (via an `out` param) the string[] of option names. Should handle an empty, single option, and multi-option-containing string. Should also trim whitespace from each option name found. (4) Write test cases for that function too.

lhmcgann commented on tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchySpecTests.cs:
```diff
@@ -0,0 +1,552 @@
+using EStimLibrary.Extensions.SpatialModel.StringHierarchy;
+
+
+namespace EStimLibrary.UnitTests.Extensions.SpatialModel.StringHierarchy;
```
I changed it here, but calling attention to it so you know in future: use filscoped namespace syntax, i.e., just put a ; at the end of the namespace line so you don't need the extra {} around the whole body of the file and don't need to indent that extra level.

lhmcgann commented on tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchySpecTests.cs:
```diff
@@ -0,0 +1,552 @@
+    [InlineData("region1 | modifier1",
+                new[] { "region1" },
+                new[] { "modifier1" })]
+    public void ConstructorFullSpec_ValidInputs_ShouldInit(string fullSpec, 
```
Try to have test functions follow the format described in the "Our Coding Practices" page in our OneNote: <CLassMethoBeingTested>_<Scenario>_Should<ExpectedOutcome>. I know I also have the note in there you can exclude the <Scenario> bit if using [Theory] tags since there are multiple specific scenarios, but if all those specific scenarios fit into a more general scenario, e.g., "ValidInputs" like you kinda already have in you "Should<...>" section, just make that th <Scenario> piece.

lhmcgann commented on tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchySpecTests.cs:
```diff
@@ -0,0 +1,552 @@
+                new[] { "region1" },
+                new[] { "modifier1" })]
+    public void ConstructorFullSpec_ValidInputs_ShouldInit(string fullSpec, 
+        string[] expectedRegions, string[] expectedModifiers)
```
I'm also getting 2 failed test cases for this test method. Should be a simple fix of making the expected values lowercase

lhmcgann commented on tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchySpecTests.cs:
```diff
@@ -0,0 +1,552 @@
+    }
+
+    /// <summary>
+    /// Test the constructor StringHierarchySpec(string fullSpec), where the
```
Other style things I started fixing but will let you finish (Ik nit-picky and maybe tedious, but I'm being annoying about it now so we get in the habit of doing it in the first place): try to have lines be <= 80 characters and enter/indent after that. This is to help readability.

If [InlineData()] cases are multi-line too, just put the comments in a line on top of them rather than at the end of the last line. That way can read top-down and know what's coming.

lhmcgann commented on tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchySpecTests.cs:
```diff
@@ -0,0 +1,552 @@
+    [InlineData(new[] { "RegionA", "RegionB" }, new[] { "Modifier1" }, "RegionA, RegionB | Modifier1")]
+    [InlineData(new[] { "Main" }, new string[] { }, "Main")]
+    [InlineData(new[] { "X", "Y", "Z" }, new[] { "North", "South" }, "X, Y, Z | North, South")]
+    public void ToString_ShouldReturnFullSpecRepresentation(
```
All internal representations should be stored as lowercase since they go through the Parse...() methods when the object is constructed. --> Why are these uppercase "expected" values succeeding? --> May be a bug somewhere that needs fixing

lhmcgann commented on src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchySpec.cs:
```diff
@@ -81,118 +117,70 @@ public static string JoinFullSpec(string[] regionSet, string[] modifierSet)
+        return true;
     }
 
     public virtual bool Equals(StringHierarchySpec? other)
```
Check the coverage report for Equals() and GetHashCode(). I think just a few test cases missing to do with when "other" is null.

lhmcgann commented on src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchySpec.cs:
```diff
@@ -16,20 +16,38 @@ public record StringHierarchySpec(string[] RegionSet, string[] ModifierSet)
         $"region1{REGIONS_DELIMITER}no-option-region2{REGIONS_DELIMITER}..." +
         $"{REGIONS_MODIFIERS_DELIMITER}modifier{MODIFIERS_DELIMITER}...";
 
+    /// <summary>
```
Add doc comments to:
- overall class
- properties
- methods

lhmcgann commented on tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchySpecTests.cs:
```diff
@@ -0,0 +1,552 @@
+    [InlineData("region1 | modifier1, modifier2, modifier3",
+                new[] { "region1" },
+                new[] { "modifier1", "modifier2", "modifier3" })] // Single region with multiple modifiers
+    public void ParseFullSpec_ShouldParseWithExpectedInputs(string fullSpec, string[] expectedRegions, string[] expectedModifiers)
```
Test cases to add:
- regions with options: different numbers of options, different regions in the path with or without options
- extra whitespace, e.g., before commas (I think you have this already), between option and region base names
- also some expected to fail test cases, e.g., when there are multiple | delimiters or other "dumb user behavior" examples --> what is the expected system behavior given those inputs? are they handled accordingly? "failing" here can also mean parsing the spec semantically incorrectly (i.e., not what the user meant) but what is syntactically correct

johnshi314 committed:
added TryParseOptions(), fixed function names to meet convention, made lines shorter, ensure internal representations are lowercase, added null tests for Equals() and GetHashCode()additional tests for ParseFullSpec()

johnshi314 committed:
added TryParseOptions(), fixed function names to meet convention, made lines shorter, ensure internal representations are lowercase, added null tests for Equals() and GetHashCode()additional tests for ParseFullSpec()

lhmcgann committed:
Robust spec parsing and testing

- null/empty handling of spec inputs
- lower-casing and delimeter whitespace consistency
- full coverage for test cases

lhmcgann committed:
Doc comments

lhmcgann committed:
Merge branch 'release-v2.0' into StringHierarchySpec_Testing

lhmcgann committed:
Region fixes to handle overlapping modifiers and tests for them

lhmcgann committed:
Fixed area, location, and region behavior to match robust spec parsing and null region rejection

lhmcgann committed:
Fixed indentation

lhmcgann committed:
Added spec equals test case

lhmcgann committed:
Comments explaining code non-coverage
